[yjunyeu] iP

yjunyeu opened the pull request:
# Waty Task Manager

**Waty** frees your mind of having to remember things you need to do. It's:

- **text-based**
- *easy to learn*
- **~FAST~ SUPER FAST** to use

All you need to do is:

1. **Download** it from [here](https://github.com/yjunyeu/ip/releases/tag/v0.1).
2. **Double-click** it.
3. **Add** your tasks.
4. Let it **manage your tasks** for you. 😉

And it is **FREE!**

> *“Your mind is for having ideas, not holding them.”* – David Allen ([source](https://dansilvestre.com/productivity-quotes))

### Features:
- Managing tasks
- Managing deadlines (coming soon)
- Reminders (coming soon)

If you are a **Java programmer**, you can use it to practice Java too. Here's the main method:

```java
public class Waty {
    public static void main(String[] args) {
        new Waty().run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yjunyeu committed:
Rename Duke.java to Waty.java

yjunyeu committed:
Add echo functionality to chatbot

yjunyeu committed:
Add horizontalLineFormatter function to simplify chatbot replies

yjunyeu committed:
Add the ability to store tasks and list tasks by the user

yjunyeu committed:
Create a new Task class in Task.java to represent tasks

yjunyeu committed:
Add the ability to mark tasks as done and also unmark tasks

yjunyeu committed:
Add ToDo, Deadline and Event classs

yjunyeu committed:
Add support for multiple task types

yjunyeu committed:
Add automated UI testing

yjunyeu committed:
Add InvalidArgumentException and InvalidCommandException to handle different exceptions

yjunyeu committed:
Add support for deleting tasks from the list

yjunyeu committed:
Add getSaveData() method for serialization

yjunyeu committed:
Add loadTask() static methods

yjunyeu committed:
Add save functionality

yjunyeu committed:
Merge branch 'branch-Level-7'

yjunyeu committed:
Add support for date and time handling in tasks

yjunyeu committed:
Merge branch 'branch-Level-8'

yjunyeu committed:
Add Ui class to handle user interactions

yjunyeu committed:
Add storage class

yjunyeu committed:
Add parser class to parse user input

yjunyeu committed:
Add command classes

yjunyeu committed:
Add TaskList class to manage tasks

yjunyeu committed:
Merge branch 'branch-moreOOP'

yjunyeu committed:
Organise classes into packages

yjunyeu committed:
Merge branch 'add-gradle-support'

yjunyeu committed:
Setup Gradle

yjunyeu committed:
Add test files for Task package

yjunyeu committed:
Add tests for storage

yjunyeu committed:
Add tests for Parser

yjunyeu committed:
Add JavaDoc comments to all classes and methods

yjunyeu committed:
Apply coding standards to all classes

yjunyeu committed:
Add find command to search tasks by keyword

yjunyeu committed:
Add JUnit tests for find command

yjunyeu committed:
Merge branch 'branch-A-JavaDoc'

yjunyeu committed:
Merge branch 'branch-A-CodingStandard'

yjunyeu committed:
Merge branch 'branch-Level-9'

Billy026 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,27 @@
+     *
+     * @return true if the application should continue running, false otherwise.
+     */
+    public boolean isRunning() { return true; }
```
Perhaps put the code in the {} brackets in new lines?

Billy026 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Represents an abstract command that can be executed.
+ */
+abstract public class Command {
```
I like how you abstracted the commands. I feel like most students would lump all commands together in one file.

Billy026 commented on src/main/java/commands/ExitCommand.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return false, signaling the application should exit.
+     */
+    public boolean isRunning() { return false; }
```
Perhaps you can split the commands into SystemCommand and UserCommand? This way you can have the abstract method isRunning only in SystemCommand.

Billy026 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Sets up the storage directory and file if they do not exist.
+     */
+    private void setupStorage() {
```
Perhaps the u in setup should be capitalised?

Billy026 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+public class Deadline extends Task {
+    private final LocalDateTime by;
+
+    /**
+     * Constructs a Deadline task.
+     * @param description
+     * @param by
+     */
+
```
Should this header comment by complete?

Billy026 reviewed (commented):
I really like how much you've abstracted everything into clear, concise classes. There's just a few minor problems with JavaDoc styling. Really glad to have a look through your coding style!

yjunyeu committed:
Fix coding standard errors

yjunyeu committed:
Add prototype gui

yjunyeu committed:
Fix bugs and add javadoc

yjunyeu committed:
Merge branch 'branch-Level-10'

yjunyeu committed:
Add varargs to find feature

mhaikelll commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+            }
+            storageWriter.close();
+        } catch (IOException e) {
+            ui.displayError(e.getMessage());
```
Perhaps can try to catch more errors in addition to IOException such as FileNotFoundException for example.

mhaikelll commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,60 @@
+        super(description);
+        this.from = LocalDateTime.parse(from, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+        this.to = LocalDateTime.parse(to, DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+    }
```
Perhaps can separate the parsing of LocalDateTime into a separate class since there seems to be a bit of repetition.

mhaikelll commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public String getSaveData() {
+        return (this.status ? "1 | " : "0 | ") + this.description;
+    }
```
Perhaps can indent the ? operator such that it follows coding standards.

mhaikelll commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Represents an abstract command that can be executed.
+ */
+public abstract class Command {
```
The use of a command abstract class is great in keeping the code neat. Well done!

mhaikelll reviewed (commented):
The code looks good overall and is very well organised. Keep up the good work!

zechary28 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,102 @@
+import commands.MarkCommand;
+import commands.UnmarkCommand;
+import exceptions.InvalidArgumentException;
+import exceptions.InvalidCommandException;
```
good job importing all of them one by one haha, convention is good

zechary28 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+ * Handles the storage and retrieval of tasks from a text file.
+ */
+public class Storage {
+    private final Path storageDir = Paths.get(System.getProperty("user.dir"), "data");
+    private final Path storageFile = storageDir.resolve("waty.txt");
```
I feel like these could be in caps since its final eg STORAGE_DIRECTORY

zechary28 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+ * Represents a task with a deadline
+ */
+public class Deadline extends Task {
+    private final LocalDateTime by;
```
variable name could be confusing since 'by' is a preposition and not a noun, maybe dueDate would be clearer

zechary28 reviewed (commented):
Generally lgtm except for few nits. nice job!

yjunyeu committed:
Add assertions to validate program assumptions

Added assertions to ensure key assumptions hold throughout the
program, improving reliability and making debugging easier. These
assertions verify critical conditions such as:

- Non-null user input in Parser to prevent unexpected errors.
- Valid task indices in TaskList to avoid IndexOutOfBoundsException.
- Correct storage setup in Storage to prevent data loss.
- Properly formatted task data when loading from storage.

These assertions help detect programmer errors early and improve
code maintainability while ensuring expected behavior during
runtime.

yjunyeu committed:
Improve code quality

Refactored the codebase to enhance readability and maintainability:

- Removed the unused `main()` method in Waty.java, as it was
  redundant and no longer contributed to functionality.
- Renamed the variable `by` to `dueDate` in Deadline.java to
  better reflect its purpose, improving code clarity.

These changes reduce clutter, improve readability, and align
the code with best practices for maintainability.

yjunyeu committed:
Merge pull request #1 from yjunyeu/branch-A-Assertions

Add assertions to validate program assumptions

yjunyeu committed:
Merge pull request #2 from yjunyeu/branch-A-CodeQuality

Improve code quality

yjunyeu committed:
Fix `findTasks`: Prevent duplicate matches

The `findTasks` method allows duplicate entries when multiple
keywords match the same task. This leads to cluttered and confusing
search results.

A unique search result list improves readability and usability.

Let's,
* use a `Set<Integer>` to track matched task indices
* prevent duplicate additions to the result list
* ensure each matching task appears only once

yjunyeu committed:
Add extension to sort by date

yjunyeu committed:
Merge pull request #3 from yjunyeu/branch-C-Sort

Branch c sort

yjunyeu committed:
Fix boolean naming problem for Task

yjunyeu committed:
Fix bugs and add more error handling

yjunyeu committed:
Merge branch 'branch-A-MoreErrorHandling'

yjunyeu committed:
Add coloured bubbles for messages

yjunyeu committed:
Merge branch 'branch-A-BetterGui'

yjunyeu committed:
Add user guide

yjunyeu committed:
Format readme table

yjunyeu committed:
Fix readme table

yjunyeu committed:
Update readme table to html

yjunyeu committed:
Fix test cases and clean up code

yjunyeu committed:
Update README.md

yjunyeu committed:
Update image path in README.md

yjunyeu committed:
Fix typo in README.md

yjunyeu committed:
Fix more bugs
